### PR TITLE
Fixes for trainer with update_on_kvstore=False

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Explicitly throw warning message on unsupported cases where `update_on_kvstore` is set to False. Previously, `self._update_on_kvstore` would be silently set to True even user constructed Trainer with `update_on_kvstore=False`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/trainer.py [hunk 1]
```diff
@@ -28,6 +28,13 @@ class Trainer(object):
     """Applies an `Optimizer` on a set of Parameters. Trainer should
     be used together with `autograd`.
 
+    .. note::
+
+        `update_on_kvstore=False` is not supported in the following cases:
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply python/mxnet/gluon/trainer.py]: Thanks! I'll update the documentation. Does the rest of the changes look good?
